src/main/java/com/imsweb/mph/mprules/MpRuleInsituAfterInvasiveSameSide.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2019 Information Management Services, Inc.
+ */
+package com.imsweb.mph.mprules;
+
+import java.util.Collections;
+
+import com.imsweb.mph.MphComputeOptions;
+import com.imsweb.mph.MphConstants;
+import com.imsweb.mph.MphInput;
+import com.imsweb.mph.MphRule;
+import com.imsweb.mph.MphUtils;
+import com.imsweb.mph.MphUtils.MpResult;
+import com.imsweb.mph.internal.TempRuleResult;
+import com.imsweb.mph.mpgroups.GroupUtility;
+
+public class MpRuleInsituAfterInvasiveSameSide extends MphRule {
+
+    public MpRuleInsituAfterInvasiveSameSide(String groupId, String step) {
+        super(groupId, step);
+        setQuestion("Is there an in situ tumor following an invasive tumor?");
+        setReason("An in situ tumor diagnosed following an invasive tumor is a single primary.");
+    }
+
+    @Override
+    public TempRuleResult apply(MphInput i1, MphInput i2, MphComputeOptions options) {
+        TempRuleResult result = new TempRuleResult();
+        String beh1 = i1.getBehavior(), beh2 = i2.getBehavior();
+        if (!GroupUtility.areOppositeSides(i1.getLaterality(), i2.getLaterality()) && GroupUtility.differentCategory(beh1, beh2, Collections.singletonList(MphConstants.INSITU),
+                Collections.singletonList(MphConstants.MALIGNANT))) {
+            int latestDx = GroupUtility.compareDxDate(i1, i2);
+            //if invasive is after insitu, skip
+            if ((1 == latestDx && MphConstants.MALIGNANT.equals(beh1)) || (2 == latestDx && MphConstants.MALIGNANT.equals(beh2)))
+                return result;
+            if (!GroupUtility.validPairedSiteLaterality(i1.getLaterality(), i2.getLaterality())) {
+                result.setPotentialResult(MphUtils.MpResult.SINGLE_PRIMARY);
+                if (latestDx == -1)
+                    result.setMessageUnknownLatAndDate(this.getStep(), this.getGroupId());
+                else
+                    result.setMessageUnknownLaterality(this.getStep(), this.getGroupId());
+            }
+            else if (-1 == latestDx) {
+                result.setPotentialResult(MpResult.SINGLE_PRIMARY);
+                result.setMessageUnknownDiagnosisDate(this.getStep(), this.getGroupId());
+            }
+            else if ((1 == latestDx && MphConstants.INSITU.equals(beh1)) || (2 == latestDx && MphConstants.INSITU.equals(beh2)))
```

Comment:
You call "GroupUtility.compareDxDate()" and put the results in latestDx.  Usually "compare" methods return negative (before), 0 (equals) or positive (after). There is usually no meaning specific values other than 0. 

I see here you are comparing to -1, 1 and 2. The last one is very weird. What does that mean for the "compareDxDate" to return 2? 

If the method doesn't really behave like a regular compare, then could you use constants for those -1, 1 and 2 values so this code that uses it becomes more readable?  I didn't look at the content of that compare method and so I am not entirely sure what it does, but I think either you don't give a meaning to the returned values (other than negative, 0 and positive), or you use constants that make it clear what a "2" means.

Reply:
I moved this from the old code (the 2007 rules). I used -1 when it is unable to determine. If I wrote that again, I would probably come up with a better way.

Reply:
So "2" means that the dates are 2 years apart? 1 means 1 year apart? -1 means unable to determine. What would a 0 mean? Same year? Same date?

Reply:
This is the compare method, 1 is the first tumor is the latest and 2 means the second one is the latest. For the verify it is only -1, 0 and 1. -1 is invalid, 0 is within the given time, 1 is after the given time.

Reply:
That really not consistent with the usual "compare" methods that return negative/0/positive.

I am not asking you to change the methods or their results at all, but could you use constants at least.

DX_COMPARE_FIRST_EARLIEST =1
DX_COMPARE_FIRST_LATEST = 2
DX_COMPARE_UNABLE_TO_DETERMINE = -1

Or if you assume it's always from the point of view of the first one:

DX_COMPARE_EARLIEST =1
DX_COMPARE_LATEST = 2
DX_COMPARE_UNABLE_TO_DETERMINE = -1

Same for verify:

YEAR_VERIFY_INVALID = -1
YEAR_VERIFY_WITHIN = 0
YEAR_VERIFY_OUTSIDE = 1

Or something like that.

But just having magic numbers that all have a specific (and different) meaning is not good practice.